Cassandra 19869

ifesdjeen opened the pull request.

ifesdjeen committed:
Implement command and CFK reloading for BurnTest

ifesdjeen committed:
Simplify reloading code

ifesdjeen committed:
Revert "Simplify reloading code"

This reverts commit dfdb82a542bb30077a8893f7f382a5d14e33bbfc.

ifesdjeen committed:
Implement state cleanup

belliottsmith committed:
suggestions

belliottsmith commented on accord-core/src/main/java/accord/local/cfk/SafeCommandsForKey.java:
```diff
@@ -55,12 +55,13 @@ void updatePruned(SafeCommandStore safeStore, Command nextCommand, NotifySink no
     private void update(SafeCommandStore safeStore, @Nullable Command command, CommandsForKey prevCfk, CommandsForKeyUpdate updateCfk)
     {
-        update(safeStore, command, prevCfk, updateCfk, NotifySink.DefaultNotifySink.INSTANCE);
+        update(safeStore, command, prevCfk, updateCfk, safeStore.replay() ? NotifySink.NoOpSink.INSTANCE : NotifySink.DefaultNotifySink.INSTANCE);
```
I think we can switch behaviour more cleanly here, but also we need to either capture the notifications OR we need to actually submit them OR we need to avoid updating unmanaged altogether and batch notify them at the end - otherwise they may never be notified.

ifesdjeen replied:
Made it that we do submit notifications for now. The problem was that notifications were causing submission of new tasks, which was leading to some exceptions. Should be clean by now thanks to your patches.

belliottsmith commented on accord-core/src/main/java/accord/local/Status.java:
```diff
@@ -901,7 +901,7 @@ public static Durability nonNullOrMerge(@Nullable Durability a, @Nullable Durabi
 
         public static Durability merge(Durability a, Durability b)
```
rename or introduce `nonNullOrMerge`?

ifesdjeen replied:
I think we do not need null checks there. Removed.

belliottsmith commented on accord-core/src/main/java/accord/local/Status.java:
```diff
@@ -913,7 +913,7 @@ public static Durability merge(Durability a, Durability b)
 
         public static Durability mergeAtLeast(Durability a, Durability b)
```
rename or introduce `nonNullOrMergeAtLeast`?

ifesdjeen replied:
I think we do not need null checks there. Removed.

belliottsmith commented on accord-core/src/main/java/accord/local/Commands.java:
```diff
@@ -868,20 +903,15 @@ public static Command purge(SafeCommandStore safeStore, SafeCommand safeCommand,
                 Invariants.checkState(command.saveStatus().compareTo(TruncatedApply) < 0);
-                Invariants.checkState(command.hasBeen(PreApplied));
-                result = truncatedApply(command, Route.tryCastToFullRoute(maybeFullRoute));
+                if (!command.hasBeen(PreCommitted)) result = Command.Truncated.erasedOrInvalidOrVestigial(command);
```
I think this is a bad merge

ifesdjeen replied:
oh yes it is!

belliottsmith commented on accord-core/src/main/java/accord/local/Commands.java:
```diff
@@ -568,7 +576,25 @@ private static AsyncChain<Void> applyChain(SafeCommandStore safeStore, PreLoadCo
     }
 
-    private static void apply(SafeCommandStore safeStore, Command.Executed command)
+    public static AsyncChain<Void> applyFromReplay(SafeCommandStore safeStore, PreLoadContext context, TxnId txnId)
```
unused

belliottsmith commented on accord-core/src/test/java/accord/impl/basic/Journal.java:
```diff
@@ -49,15 +64,19 @@
+    private static final Logger logger = LoggerFactory.getLogger(Journal.class);
+
     private final Queue<RequestContext> unframedRequests = new ArrayDeque<>();
     private final LongArrayList waitForEpochs = new LongArrayList();
```
do we need this anymore?

belliottsmith committed:
suggest

belliottsmith committed:
accidental !

ifesdjeen committed:
Cleanup

belliottsmith committed:
do not reset nextInvokerId in DPL.clear

belliottsmith committed:
fix flipping Accept to AcceptedInvalidate

ifesdjeen committed:
Provide correct ballot

ifesdjeen committed:
Add forceApply to replay

ifesdjeen committed:
Avoid forcing application of truncated commands

ifesdjeen committed:
Drain all non-recurring runnables on journal replay

ifesdjeen committed:
Remove epoch waiting / delayed request processing from Journal.

ifesdjeen committed:
Review feedback

ifesdjeen committed:
More cleanup

ifesdjeen committed:
More RAM for tests

belliottsmith commented on accord-core/src/test/java/accord/impl/basic/Cluster.java:
```diff
@@ -475,11 +513,40 @@ public static Map<MessageType, Stats> run(Id[] nodes, MessageListener messageLis
             Scheduled reconfigure = sinks.recurring(configRandomizer::maybeUpdateTopology, 1, SECONDS);
+
+            Scheduled purge = sinks.recurring(() -> {
+                trace.debug("Triggering purge.");
```
It would be nice to randomise the frequency of these operations (lots of time in tests is now spent on restart, and anyway would be good to change the cycle), and also to randomise the set of nodes we shutdown for this, so that some transactions can carry on being processed (potentially) during this window, and we can exercise more interesting situations.

Or perhaps just randomise each replica independently.

ifesdjeen replied:
Added randomization.

ifesdjeen committed:
Different syntax

ifesdjeen committed:
Different syntax

ifesdjeen committed:
Fix test

belliottsmith committed:
suggestions

belliottsmith commented on accord-core/src/test/java/accord/impl/basic/Cluster.java:
```diff
@@ -219,18 +220,50 @@ public void processAll()
         checkFailures.run();
+
+        // If drain was requested, run until we have no non-recurring tasks are available
+        if (next instanceof RecurringPendingRunnable && ((RecurringPendingRunnable) next).drainAfterRunning)
```
Might it be cleaner to submit a kind of "drain" task?

belliottsmith replied:
by which I mean, instead of making the recurring task say "drain after me", the task submits a one-shot drain task whenever it wants to perform a drain?

belliottsmith replied:
This drain task could itself submit some other post-drain task to e.g. re-enable messaging (and schedule the next journal replay task)

ifesdjeen committed:
Add interval randomization. Rework drain.

ifesdjeen committed:
Simplify API for self-recurring tasks; move process non-recurring to Sinks

ifesdjeen committed:
Skip linerizability reporting

ifesdjeen committed:
Clear data in list store, too

ifesdjeen committed:
TMP: only load the last command; hardcoded seed

ifesdjeen committed:
trying

ifesdjeen committed:
Cosmetics

ifesdjeen committed:
2-phase loading

aweisberg committed:
Persists metadata syncrhonously

ifesdjeen committed:
Implement state snapshotting; improve truncation and purging

belliottsmith committed:
temporary fix for invalid use of Sets.filter

ifesdjeen committed:
Drain tasks for a specific node instead of stopping networking (for now)

ifesdjeen committed:
Unset seed

ifesdjeen committed:
Short-circuit truncated in updateRouteOrParticipants since the command can be in that state now

ifesdjeen committed:
Cleanup

ifesdjeen committed:
Fully drain _before_ application

ifesdjeen committed:
Constrain map

ifesdjeen committed:
Simplify reload code

ifesdjeen committed:
Avoid applying invalidated and truncated commands

ifesdjeen committed:
Fix a problem with not moving applied commands to appropriate state

ifesdjeen committed:
Snapshot during bootstrap

ifesdjeen committed:
Typo

ifesdjeen committed:
Add node id to journal for debugging convenience

ifesdjeen committed:
Make sure to re-count recurring

ifesdjeen committed:
Purge entries from journal entirely if they are ERASED; add a check for accidental Uninitialized

ifesdjeen committed:
Fix [16,19642,2(KW),3] is less than the most recent write timestamp [16,19925,2(KW),1] for 673:1#47225 (seed 69268820720458l)

Command has to be in Applying to qualify for force-apply, not in PreApplied, as PreApplied means we know the outcome, and we might be able to apply it. But we may not have applied preceding transactions, and may not even be able to.

ifesdjeen committed:
Fix checkstyle

ifesdjeen committed:
Revert unrelated change

ifesdjeen committed:
Fixes after rebase and an attempt to reconstruct redundant before

belliottsmith committed:
wip

belliottsmith committed:
wip: journal replay finishing touches

belliottsmith committed:
wip: journal replay finishing touches

ifesdjeen committed:
Checkstyle

ifesdjeen committed:
Make it easier to log changes in rangesForEpoch, durableBefore, redundantBefore, safeToRead, and rangesForEpoch

ifesdjeen committed:
Simplify bootstrap

ifesdjeen committed:
Make it compile

ifesdjeen committed:
Remove assert

ifesdjeen committed:
Fix build

ifesdjeen committed:
Overwrite reject before on every op

ifesdjeen committed:
Override equals for snapshot

ifesdjeen committed:
Cleanup

ifesdjeen committed:
Rework the way historical transactions are saved

ifesdjeen committed:
Rename "upsertSafeToRead", revert safe reject before

ifesdjeen committed:
Fix build

ifesdjeen committed:
Cleanup

ifesdjeen committed:
set -> unsafe ; add missing `beginAsResult`

ifesdjeen committed:
Cosmetics

ifesdjeen committed:
More cleanup

ifesdjeen committed:
Fix durable before serializer

ifesdjeen committed:
Write bootstrap at entirely
